### ISSUE TITLE
"Error! Exception while fetching hosts : [object Object]"

### ISSUE THREAD

**Issue by caio1982** [sivasamyk/logtrail#302]
I've installed Kibana 6.3.1 with Logtrail 0.1.29 and I get `Error! Exception while fetching hosts : [object Object]` with a broad red band when I go to the Logtrail page. Checking the Kibana instance I see this:

```
{"type":"response","@timestamp":"2018-08-12T18:24:55Z","tags":[],"pid":1,"method":"get","statusCode":200,"req":{"url":"/logtrail/config","method":"get","headers":{"host":"127.0.0.1:8080","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:61.0) Gecko/20100101 Firefox/61.0","accept":"application/json, text/plain, */*","accept-language":"en-US,en;q=0.5","accept-encoding":"gzip, deflate","referer":"http://127.0.0.1:8080/app/logtrail","kbn-version":"6.3.1","dnt":"1","connection":"keep-alive"},"remoteAddress":"127.0.0.1","userAgent":"127.0.0.1","referer":"http://127.0.0.1:8080/app/logtrail"},"res":{"statusCode":200,"responseTime":10,"contentLength":9},"message":"GET /logtrail/config 200 10ms - 9.0B"}
{"type":"response","@timestamp":"2018-08-12T18:24:55Z","tags":[],"pid":1,"method":"get","statusCode":304,"req":{"url":"/bundles/ebdca7741674eca4e1fadeca157f3ae6.svg","method":"get","headers":{"host":"127.0.0.1:8080","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:61.0) Gecko/20100101 Firefox/61.0","accept":"*/*","accept-language":"en-US,en;q=0.5","accept-encoding":"gzip, deflate","referer":"http://127.0.0.1:8080/bundles/commons.style.css","dnt":"1","connection":"keep-alive","if-modified-since":"Sun, 12 Aug 2018 18:12:17 GMT","if-none-match":"\"eacd5acd1258d9b09e78dbc1958744f30e38bcbd-gzip\""},"remoteAddress":"127.0.0.1","userAgent":"127.0.0.1","referer":"http://127.0.0.1:8080/bundles/commons.style.css"},"res":{"statusCode":304,"responseTime":2,"contentLength":9},"message":"GET /bundles/ebdca7741674eca4e1fadeca157f3ae6.svg 304 2ms - 9.0B"}
Error while fetching hosts TypeError: Cannot read property 'hosts' of undefined
    at /usr/share/kibana/plugins/logtrail/server/routes/server.js:248:35
    at bound (domain.js:301:14)
    at runBound (domain.js:314:12)
    at tryCatcher (/usr/share/kibana/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/usr/share/kibana/node_modules/bluebird/js/main/promise.js:503:31)
    at Promise._settlePromiseAt (/usr/share/kibana/node_modules/bluebird/js/main/promise.js:577:18)
    at Promise._settlePromises (/usr/share/kibana/node_modules/bluebird/js/main/promise.js:693:14)
    at Async._drainQueue (/usr/share/kibana/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/usr/share/kibana/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues (/usr/share/kibana/node_modules/bluebird/js/main/async.js:15:14)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
{"type":"response","@timestamp":"2018-08-12T18:24:56Z","tags":[],"pid":1,"method":"post","statusCode":200,"req":{"url":"/logtrail/hosts","method":"post","headers":{"host":"127.0.0.1:8080","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:61.0) Gecko/20100101 Firefox/61.0","accept":"application/json, text/plain, */*","accept-language":"en-US,en;q=0.5","accept-encoding":"gzip, deflate","referer":"http://127.0.0.1:8080/app/logtrail","content-type":"application/json;charset=utf-8","kbn-version":"6.3.1","content-length":"500","dnt":"1","connection":"keep-alive"},"remoteAddress":"127.0.0.1","userAgent":"127.0.0.1","referer":"http://127.0.0.1:8080/app/logtrail"},"res":{"statusCode":200,"responseTime":10,"contentLength":9},"message":"POST /logtrail/hosts 200 10ms - 9.0B"}
{"type":"response","@timestamp":"2018-08-12T18:24:56Z","tags":[],"pid":1,"method":"post","statusCode":200,"req":{"url":"/logtrail/search","method":"post","headers":{"host":"127.0.0.1:8080","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:61.0) Gecko/20100101 Firefox/61.0","accept":"application/json, text/plain, */*","accept-language":"en-US,en;q=0.5","accept-encoding":"gzip, deflate","referer":"http://127.0.0.1:8080/app/logtrail","content-type":"application/json;charset=utf-8","kbn-version":"6.3.1","content-length":"614","dnt":"1","connection":"keep-alive"},"remoteAddress":"127.0.0.1","userAgent":"127.0.0.1","referer":"http://127.0.0.1:8080/app/logtrail"},"res":{"statusCode":200,"responseTime":9,"contentLength":9},"message":"POST /logtrail/search 200 9ms - 9.0B"}
```

I am testing it on Kubernetes with the following Docker image setup (available through my Docker Hub if needed for testing):

```
FROM docker.elastic.co/kibana/kibana-oss:6.3.1
MAINTAINER caio1982
RUN ./bin/kibana-plugin install https://github.com/sivasamyk/logtrail/releases/download/v0.1.29/logtrail-6.3.1-0.1.29.zip
```

Without Logtrail the same Kibana setup works normally. Even though I see `{"type":"log","@timestamp":"2018-08-12T18:24:55Z","tags":["info","status"],"pid":1,"message":"Error while loading config from Elasticsearch. Will use local"}` in the log, all seems good and working without Logtrail.

My indices are all `logstash-*`.

Any idea, folks? :-(

**Comment by caio1982** [comment]
I get the exact same error if I try the same setup with 6.3.2 by the way, including the corresponding Logtrail 6.3.2-0.1.29 release.

**Comment by caio1982** [comment]
Ok, so it seems Logtrail likes the filebeat-* kind of index by default so it does not work with the EFK stack used by Kubernetes people, usually. What I did to fix the default behavior of Logtrail while being couple with Kibana in Kubernetes is documented here: https://hub.docker.com/r/caio1982/kibana-logtrail/

TL;DR: shipped a ConfigMap for Kubernetes to mount the Logtrail config on top of any existing default file (which also happens to have the same content in my Docker image, just in case).

Anyways... I will leave this issue open so you can evaluate, maybe, adding a more user-friendly message to the UI or even perhaps allowing the configuration to be changed via the UI itself 😅

**Comment by sivasamyk** [comment]
Thanks for the issue. Will fix the error message. Meanwhile, logtrail has an option to load `logtrail.json` from Elasticsearch so you don't need the config map to create logtrail json in the Kibana pod. Check out https://github.com/sivasamyk/logtrail/blob/master/docs/how_to.md#3-load-logtrail-configuration-from-elasticsearch

**Comment by C0rn3j** [comment]
I think I also ran into the same issue but I had a previously working logtrail on 5.x, after upgrade to 6.x (currently 6.4) I can't seem to get logtrail to work (logtrail 6.4 was installed).
https://i.imgur.com/Shuh1Cz.png

Not sure if mentioned workarounds in here would work for my situation?

```
Sep 08 12:31:51 kibana kibana[227914]: Error while fetching hosts TypeError: Cannot read property 'hosts' of undefined
Sep 08 12:31:51 kibana kibana[227914]:     at /usr/share/kibana/plugins/logtrail/server/routes/server.js:248:35
Sep 08 12:31:51 kibana kibana[227914]:     at bound (domain.js:301:14)
Sep 08 12:31:51 kibana kibana[227914]:     at runBound (domain.js:314:12)
Sep 08 12:31:51 kibana kibana[227914]:     at tryCatcher (/usr/share/kibana/node_modules/bluebird/js/main/util.js:26:23)
Sep 08 12:31:51 kibana kibana[227914]:     at Promise._settlePromiseFromHandler (/usr/share/kibana/node_modules/bluebird/js/main/promise.js:503:31)
Sep 08 12:31:51 kibana kibana[227914]:     at Promise._settlePromiseAt (/usr/share/kibana/node_modules/bluebird/js/main/promise.js:577:18)
Sep 08 12:31:51 kibana kibana[227914]:     at Promise._settlePromises (/usr/share/kibana/node_modules/bluebird/js/main/promise.js:693:14)
Sep 08 12:31:51 kibana kibana[227914]:     at Async._drainQueue (/usr/share/kibana/node_modules/bluebird/js/main/async.js:123:16)
Sep 08 12:31:51 kibana kibana[227914]:     at Async._drainQueues (/usr/share/kibana/node_modules/bluebird/js/main/async.js:133:10)
Sep 08 12:31:51 kibana kibana[227914]:     at Immediate.Async.drainQueues (/usr/share/kibana/node_modules/bluebird/js/main/async.js:15:14)
Sep 08 12:31:51 kibana kibana[227914]:     at runCallback (timers.js:810:20)
Sep 08 12:31:51 kibana kibana[227914]:     at tryOnImmediate (timers.js:768:5)
Sep 08 12:31:51 kibana kibana[227914]:     at processImmediate [as _immediateCallback] (timers.js:745:5)
```

**Comment by sivasamyk** [comment]
@C0rn3j Are running Kibana in docker/k8s setup?  Can you share your logtrail.json file?

**Comment by sivasamyk** [comment]
Thanks for the info. I see you are using the default logtrail.json. Can you check the mapping in ES for `filebeat-*` index and check if `beat.hostname.keyword` mapping is present? Can you share the result of the following query?

```sh
curl -XPOST -H "Content-Type:application/json" localhost:9200/filebeat-*/_search -d ' { "size" : 0, "aggs": { "hosts": { "terms" : { "field" : "beat.hostname.keyword" } } } }'
```

**Comment by C0rn3j** [comment]
```
{"took":0,"timed_out":false,"_shards":{"total":0,"successful":0,"skipped":0,"failed":0},"hits":{"total":0,"max_score":0.0,"hits":[]}}
```

Am a bit rusty with ES, but in Index Management for ES under Kibana it only shows logstash-*

![image](https://user-images.githubusercontent.com/1641362/45255431-a630b680-b386-11e8-8156-ee747ef87bdb.png)

**Comment by sivasamyk** [comment]
Aha. The it mean you don't have `filebeat-*` index pattern you ES. You need to change default_index in  `logtrail.json` to `logstash-*` and adjust the field mappings accordingly. Once done you can upload the logtrail.json config to ES using https://github.com/sivasamyk/logtrail/blob/master/docs/how_to.md#3-load-logtrail-configuration-from-elasticsearch (recommended in container setup.)

**Comment by C0rn3j** [comment]
Got it running, thanks!

**Comment by sivasamyk** [comment]
Adjusted error message display while fetching hosts. download the latest release from https://github.com/sivasamyk/logtrail/releases/download/v6.4.0_0.1.30/logtrail-6.4.0-0.1.30.zip
